Still more `rustc_mir_dataflow` cleanups

nnethercote opened the pull request:
A sequel to #118203, #118230, #118638, and #131481.

I'm pleased with this PR. It cleans up a few things that have been bugging me for a while. It took quite some effort to find these improvements.

r? @cjgillot

nnethercote committed:
In `MaybeRequiresStorage`, store `Results` instead of `ResultsCursor`.

This requires a `clone` call in `check_for_move`, but makes
`MaybeRequiresStorage` an immutable analysis, which enables the next
commit.

nnethercote committed:
Make analyses and visitables immutable.

Much nicer.

nnethercote committed:
Make results immutable.

Because most of them are. The one exception is `ConstAnalysis`, which
gains a `RefCell` to provide interior mutability. This is a bit
annoying, but I think it's worth inconveniencing the one analysis that
requires mutability in exchange for making the interface simpler for all
the other analyses.

nnethercote committed:
Rename `BlockFormatter::results` as `BlockFormatter::cursor`.

Because it's a `ResultsCursor`, not a `Results`.

nnethercote committed:
Simplify `graphviz::Formatter`.

`Formatter` currently has a `RefCell<Option<Results>>` field. This is so
the `Results` can be temporarily taken and put into a `ResultsCursor`
that is used by `BlockFormatter`, and then put back, which is messy.

This commit changes `Formatter` to have a `RefCell<ResultsCursor>` and
`BlockFormatter` to have a `&mut ResultsCursor`, which greatly
simplifies the code at the `Formatter`/`BlockFormatter` interaction
point in `Formatter::node_label`.

The commit also:
- documents the reason for the `RefCell`;
- adds a `Formatter::body` method, replacing the `Formatter::body`
  field.

nnethercote committed:
Return label from `write_node_label`.

Instead of mutating an empy label. Because it's conceptually simpler.

nnethercote committed:
Reduce some visibilities.

nnethercote committed:
Remove `Analysis::into_engine`.

This is a standard pattern:
```
MyAnalysis.into_engine(tcx, body).iterate_to_fixpoint()
```
In fact, `into_engine` and `iterate_to_fixpoint` are always called in
pairs, but sometimes with a builder-style `pass_name` calls between
them. But a builder-style interface is overkill here. This has been
bugging me a for a while.

This commit:
- Merges `Engine::new` and `Engine::iterate_to_fixpoint`. This removes
  the need for `Engine` to have fields, leaving it as a trivial type
  that the next commit will remove.
- Renames `Analysis::into_engine` as `Analysis::iterate_to_fixpoint`,
  gives it an extra argument for the optional pass name, and makes it
  call `Engine::iterate_to_fixpoint` instead of `Engine::new`.

This turns the pattern from above into this:
```
MyAnalysis.iterate_to_fixpoint(tcx, body, None)
```
which is shorter at every call site, and there's less plumbing required
to support it.

nnethercote committed:
Remove `Engine`.

It's no longer needed. `Engine::iterate_to_fixpoint` can be inlined into
`Analysis::iterate_to_fixpoint` and removed. The commit also renames
`engine.rs` as `results.rs`.

nnethercote committed:
Make it possible for `ResultsCursor` to borrow a `Results`.

`ResultsCursor` currently owns its `Results`. But sometimes the
`Results` is needed again afterwards. So there is
`ResultsCursor::into_results` for extracting the `Results`, which leads
to some awkwardness.

This commit adds `ResultsHandle`, a `Cow`-like type that can either
borrow or own a a `Results`. `ResultsCursor` now uses it. This is good
because some `ResultsCursor`s really want to own their `Results`, while
others just want to borrow it.

We end with with a few more lines of code, but get some nice cleanups.
- `ResultsCursor::into_results` is removed.
- `Formatter::into_results` is removed.
- `write_graphviz_results` now just borrows a `Results`, instead of the
  awkward "take ownership of a `Results` and then return it unchanged"
  pattern.
- `MaybeRequiresStorage` can borrow the `MaybeBorrowedLocals` results,
  avoiding two `clone` calls: one in `check_for_move` and one in
  `locals_live_across_suspend_points`.
- `Results` no longer needs to derive `Clone`.

This reinstates the cursor flexibility that was lost in #118230 -- which
removed the old `ResultsRefCursor` and `ResultsCloneCursor` types -- but
in a *much* simpler way. Hooray!

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9e735caf62fc717556c3cc3623c7a136b01aa515 with merge 1c188eb1f522be2697a49819f69932c2a96181ee...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9e735caf62fc717556c3cc3623c7a136b01aa515","merge_sha":"1c188eb1f522be2697a49819f69932c2a96181ee"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11489868247/job/31982567937)
Build commit: 1c188eb1f522be2697a49819f69932c2a96181ee (`1c188eb1f522be2697a49819f69932c2a96181ee`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11489868247/job/31982567937"},"merge_sha":"1c188eb1f522be2697a49819f69932c2a96181ee"} -->

rust-timer commented:
Queued 1c188eb1f522be2697a49819f69932c2a96181ee with parent 4f2f477fded0a47b21ed3f6aeddeafa5db8bf518, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4f2f477fded0a47b21ed3f6aeddeafa5db8bf518&end=1c188eb1f522be2697a49819f69932c2a96181ee).
There are currently 4 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~5.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

nnethercote committed:
Remove an unnecessary method.

Because the default method is also a no-op.

rust-timer commented:
Finished benchmarking commit (1c188eb1f522be2697a49819f69932c2a96181ee): [comparison URL](https://perf.rust-lang.org/compare.html?start=4f2f477fded0a47b21ed3f6aeddeafa5db8bf518&end=1c188eb1f522be2697a49819f69932c2a96181ee&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=4f2f477fded0a47b21ed3f6aeddeafa5db8bf518&end=1c188eb1f522be2697a49819f69932c2a96181ee&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.4%, 1.0%]   | 7     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 2     |
| All ❌✅ (primary)                 | 0.2%  | [0.2%, 0.2%]   | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4f2f477fded0a47b21ed3f6aeddeafa5db8bf518&end=1c188eb1f522be2697a49819f69932c2a96181ee&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=4f2f477fded0a47b21ed3f6aeddeafa5db8bf518&end=1c188eb1f522be2697a49819f69932c2a96181ee&stat=cycles:u)
<details>
<summary>Results (primary -2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.2% | [-2.2%, -2.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.2% | [-2.2%, -2.2%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4f2f477fded0a47b21ed3f6aeddeafa5db8bf518&end=1c188eb1f522be2697a49819f69932c2a96181ee&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 781.235s -> 781.882s (0.08%)
**Artifact size**: 333.60 MiB -> 333.62 MiB (0.01%)

nnethercote commented:
The `await-call-tree` regression looks like it might be real. I ran Cachegrind locally but it wasn't informative. A bunch of functions got more instructions, a bunch got fewer, and there was no clear pattern. The regression is so small and only on a stress test so I don't think it's important.

@rustbot label: +perf-regression-triaged

nnethercote committed:
Remove some more unnecessary `mut` usage.

cjgillot commented on compiler/rustc_mir_dataflow/src/impls/storage_liveness.rs:
```diff
@@ -283,15 +282,15 @@ impl<'tcx> Analysis<'tcx> for MaybeRequiresStorage<'_, 'tcx> {
+    fn check_for_move(&self, trans: &mut <Self as Analysis<'tcx>>::Domain, loc: Location) {
+        let borrowed_locals = self.borrowed_locals.clone().into_results_cursor(self.body);
+        let mut visitor = MoveVisitor { trans, borrowed_locals };
+        visitor.visit_location(self.body, loc);
```
This changes the complexity of the analysis. The former code use a cursor, which remembers the current position. The latter code recreates a cursor, which implicitly needs to recompute results from bb0. Could we use a `RefCell` or something?

nnethercote replied:
Hmm, ok. I've reinstated the cursor within a `RefCell`.

nnethercote committed:
Reinstate `ResultsCursor` in `MaybeRequiresStorage`.

Because creating a new cursor in every `check_for_move` call might
results in quadratic behaviour. The cursor must be wrapped in a
`RefCell`, now that analyses are passed everywhere by `&` instead of
`&mut`.

nnethercote committed:
Tweak `MaybeBorrowedLocals::transfer_function` usage.

First, remove the `self` param, which is unnecessary.

Also, in `MaybeRequiresStorage::apply_before_statement_effect`, call
`transfer_function` directly, as is already done in
`MaybeRequiresStorage::apply_before_terminator_effect`. This makes it
clear that the operation doesn't rely on the `MaybeBorrowedLocals`
results.

nnethercote commented:
I have added two new commits: one that reinstates the cursor within `MaybeRequireStorage`, and one that does minor cleanups.

nnethercote commented:
@Jarcho talked on Zulip about needing analysis mutability for some Clippy analyses. So I will close this PR and redo the non-mutability bits in other PRs.
